src/levanter/models/attention.py

```diff
@@ -57,27 +59,77 @@ def dot_product_attention(
         raise ValueError("QPos and KPos must be different")
 
     if use_flash:
-        from levanter.models.flash_attention import BLOCK_SIZE, flash_attention
-
-        if flash_block_size is None:
-            flash_block_size = BLOCK_SIZE
-
-        return flash_attention(
-            QPos,
-            KPos,
-            Key,
-            query,
-            key,
-            value,
-            block_size=flash_block_size,
-            mask=mask,
-            bias=bias,
-            dropout=dropout,
-            inference=inference,
-            key=prng,
-            dtype=attention_dtype,
-            precision=precision,
-        )
+        accelerator_type = xla_bridge.get_backend().platform.lower()
+
+        if accelerator_type == "tpu" or accelerator_type == "cpu":
+            from levanter.models.flash_attention import flash_attention
+
+            if flash_block_size is None:
+                seq_len = query.shape["position"]
+                flash_block_size = seq_len
+
+            return flash_attention(
+                QPos,
+                KPos,
+                Key,
+                query,
+                key,
+                value,
+                block_size=flash_block_size,
+                mask=mask,
+                bias=bias,
+                dropout=dropout,
+                inference=inference,
+                key=prng,
+                dtype=attention_dtype,
+                precision=precision,
+            )
+        elif accelerator_type == "gpu":
+            from transformer_engine.jax.fused_attn import AttnBiasType, AttnMaskType, self_fused_attn
+
+            # TransformerEngine looks for qkv shape:
+            # batch_shape, max_seqlen, nqkv, num_heads, head_dim; where nqkv = 3
+            q_ = haliax.rearrange(query, ("batch", "position", "heads", "head_size")).array
+            k_ = haliax.rearrange(key, ("batch", "key_position", "heads", "head_size")).array
+            v_ = haliax.rearrange(value, ("batch", "key_position", "heads", "head_size")).array
+
+            qkv = jnp.stack((q_, k_, v_), axis=2)
```

Comment:
this mode doesn't work for cross attention or GQA. Could we detect that (the number of heads is different) and do the right thing? at least for GQA? this is what flax does: https://github.com/NVIDIA/TransformerEngine/blob/a174985b91db023181465a3ec9ad161a68787c57/transformer_engine/jax/flax/transformer.py#L742-L751

probably equally easy to check cross and GQA: just check if either size differs